### PR TITLE
Support undefined size for grid

### PR DESCRIPTION
**Pull request by pleku**
Part of #966

### CODE REVIEW

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataView.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.vaadin.flow.data.provider.LazyDataView;
+
+/**
+ * Interface for lazy data view for grid.
+ *
+ * @param <T>
+ *            the type of the items in grid
+ */
+public interface GridLazyDataView<T> extends GridDataView<T>, LazyDataView<T> {
```
Comment (caalador) [inline vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataView.java:27]: Do we need the empty interface or could it be a class as GridListDataView and just implement the interfaces directly.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataViewImpl.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+     * @param component
+     *            the grid component
+     */
+    public GridLazyDataViewImpl(DataCommunicator<T> dataCommunicator,
+            Grid<T> component) {
+        super(dataCommunicator::getDataProvider, component);
+        this.dataCommunicator = dataCommunicator;
+    }
+
+    private DataCommunicator<T> getDataCommunicator() {
```
Comment (mshabarov) [inline vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataViewImpl.java:52]: `DataCommunicator` is going to be used for all lazy data view implementations, isn't it? If so, we could create AbstractLazyDataView and put the common methods there.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataView.java (hunk 1 above)
Comment (mshabarov) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataView.java:27]: +1

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 3]
```diff
@@ -2417,6 +2421,11 @@ public void setItems(Stream<T> streamOfItems) {
         return dataCommunicator;
     }
 
+    @Override
+    public GridLazyDataView<T> getLazyDataView() {
+        return new GridLazyDataViewImpl(getDataCommunicator(), this);
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:2426]: +1

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractUndefinedSizeGridPage.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                + query.getPreviousSizeEstimate() + ", "
+                + query.getRequestedRangeEnd());
+        logPanel.addComponentAsFirst(log);
+        if (sizeCallbackEstimate < 1) {
+            // using +20% as the default
+            int previousSizeEstimate = query.getPreviousSizeEstimate();
+            if (previousSizeEstimate == 0) {
+                return DEFAULT_INITIAL_SIZE_ESTIMATE;
+            }
+            return new BigDecimal(previousSizeEstimate)
```
Comment (mshabarov) [inline vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractUndefinedSizeGridPage.java:252]: We discussed it briefly before, but anyway I would ask why you use BigDecimal here? Don't we enough with a precision provided by `double`? It's a just an estimation which shouldn't be as much precise as possible. But, if we assume that we could use both `double` and `BigDecimal` here, and it's just an example for developers and the precision choice is up to them, then it's fine.

File: vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/grid/demo/GridDemo.java [hunk 5]
```diff
@@ -684,53 +690,120 @@ private void createBasicUsage() {
...
+         * provided.
+         */
+        GridLazyDataView<Person> lazyDataView = grid.setDataSource(
+                query -> personService
+                        .fetch(query.getOffset(), query.getLimit()).stream(),
+                query -> personService.count());
+
+        // The grid can be on switched back to undefined size through the
+        // API in the lazy data view:
+        // lazyDataView.setDefinedSize(false);
```
[the remark this answers is not in the document]
Comment (pleku) [reply vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/grid/demo/GridDemo.java]: Done

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataViewImpl.java (hunk 2 above)
Comment (pleku) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridLazyDataViewImpl.java]: Yes like I said already earlier, should be done. It is now there

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pleku) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java]: Yeah not done on purpose as discussed in other PRs.

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractUndefinedSizeGridPage.java (hunk 4 above)
Comment (pleku) [reply vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractUndefinedSizeGridPage.java:252]: This is a test class, what difference does it make ? At least it is consistent

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (caalador) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java]: Yes. But should throw a better exception than the validation exception here pointing the developer in the correct direction (getDataView/getListDataView) as the validation message is quite generic.

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java [hunk 6]
```diff
@@ -59,7 +59,7 @@ public TreeGridBasicFeaturesPage() {
         grid.setSelectionMode(SelectionMode.SINGLE);
         grid.setColumns("id", HierarchicalTestBean::toString,
                 Arrays.asList("id", "depth", "index"));
-        grid.setDataProvider(new LazyHierarchicalDataProvider(3, 2));
+        grid.setDataSource(new LazyHierarchicalDataProvider(3, 2));
```
Comment (caalador) [inline vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java:62]: These should stay as `setDataProvider` as tree grid doesn't have a good DataView implementation and not using setDataSource in the test pages would give a better indication on that.

Comment (caalador) [reply vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java:62]: Same for TreeGridDetachAttachPage, TreeGridRefreshAllPage and TreeGridScrollingPage.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pleku) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java]: The message is changed in the Flow PR for `AbstractDataView` to
```
                    "%s only supports '%s' or it's subclasses, but was given a '%s'."
                            + "%nUse either 'getLazyDataView()', 'getListDataView()'"
                            + " or getDataView() according to the used data type.",
```

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java (hunk 6 above)
Comment (pleku) [reply vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java:62]: Done

**Pull request merged**